rfcs/0006-configuration-constants.md

```diff
@@ -0,0 +1,128 @@
+
+The configuration variables for a program can be identified by the `configurable` keyword within the Sway language. This denotes a constant within the program that has no known value at compile time. The value will be provided at configuration time by the SDK or some other consumer of the configuration-time-constant descriptor. As an example:
+
+```rust
+// main.sw
+script;
+
+configurable {
+    CONTRACT_ADDRESS: b256,
+    PRICE_RATIO: u64,
+}
+
+fn main() { ... }
+```
+```
+// configurable-constants.json
+
+{
+    "CONTRACT_ADDRESS": {
+        "type": "b256",
+        "offset": "12345"
+    }
+}
+```
+
+If you are not using the SDK, it will be possible to define these values by specifying a `constants.sw` flag. This will work with the [notion of contract dependencies](https://github.com/FuelLabs/sway-rfcs/pull/19#issuecomment-1286351753). This Sway file must export the required constants as specified by the `configurable` block.
+
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The `type` field of the JSON file should describe the type of the configurable in the same manner as types are described in the JSON ABI file.
+
+No two configurables in a program should have the same name to prevent collisions.
+
+The range from the offset to the offset plus the size of the type will represent bytes within the bytecode that will be overwritten by the SDK, which will write values according to the ABI encoder's memory layout, which is consistent with Sway's memory layout.
+
+The change is not breaking as it is entirely new functionality.
+
+Configuration time constants may not be defined in any library code and must be defined at the top level. Should a library need to reference a configurable value, it should utilize design patterns such as the [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) to ingest the constants from the top level.
+
+The `pub` keyword will operate as normal: if it is `pub`, then it is public and importable from elsewhere in the Sway program. If not, then it is only referencable in the scope it is defined within.
+
+## Deploying with and without the SDK
+
+With the SDK, the deployment and assignment of values will all be handled within the SDK's API. Without the SDK, `forc` will accept command line arguments to specify these values, or allow for their specification in the manifest. In the future, we may wish to introduce more methods of defining these values, like environment variables, but it will not be necessary for the MVP. If the values are specified in more than one way, the SDK would take precedence. This is because the SDK does not actually compile the bytecode in this process; it merely overwrites the indices. This is final and precedence cannot be specified by the user.
```

Comment:
> Without the SDK, `forc` will accept command line arguments to specify these values, or allow for their specification in the manifest.

That's not how we're going to do it anymore, right? We're going to go with `constants.sw` approach?

Reply:
Good catch! Agreed, I don't think we should make any promises about offering command line arguments in `forc` here. I made a suggestion [here](https://github.com/FuelLabs/sway-rfcs/pull/19#discussion_r1003867760) that clarifies why it's needed for `forc` and the proposed behaviour.